File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
#include<iostream>
using namespace std;

int Bfs(vector<vector<int>>& maps, vector<vector<bool>>& visited)
{
    int cnt = 1;
    
    queue<pair<int,int>> q;
    
    q.push({0, 0});
    visited[0][0] = true;
    
    vector<int> dy = {-1, 0, 1, 0};
    vector<int> dx = {0, -1, 0, 1};
    
    while(!q.empty())
    {
        int size = q.size();
        
        for(int k = 0; k < size; k++)
        {
            auto [y, x] = q.front();
            q.pop();
            
            for(int i = 0; i < 4; i++)
            {
                int new_y = y + dy[i];
                int new_x = x + dx[i];
                
                if(new_y < 0 || new_y >= maps.size() || new_x < 0 || new_x >= maps[0].size()) continue;
                if(maps[new_y][new_x] == 0) continue;
                if(visited[new_y][new_x]) continue;
                
                if(new_y == maps.size()-1 && new_x == maps[0].size()-1) return cnt+1;
                
                visited[new_y][new_x] = true;
                q.push({new_y, new_x});
            }
        }
        
        cnt++;
    }
    
    return -1;
}

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    
    vector<vector<bool>> visited(maps.size(), vector<bool>(maps[0].size(), false));
    
    answer = Bfs(maps, visited);    
    
    return answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cs
using System.Collections.Generic;

class Solution {
    int[] dy = {-1, 0, 1, 0};
    int[] dx = {0, 1, 0, -1};
    
    public int solution(int[,] maps) {
        int answer = -1;
        
        int N = maps.GetLength(0);
        int M = maps.GetLength(1);
        
        bool[,] visited = new bool[N,M];
        
        Queue<(int, int)> q = new Queue<(int, int)>();
        q.Enqueue((0, 0));
        visited[0,0] = true;
        
        int depth = 1;
        while(q.Count > 0)
        {
            int size = q.Count;
            
            for(int t = 0; t < size; t++)
            {
                var (y, x) = q.Dequeue();
            
                for(int i = 0 ; i < 4; i ++)
                {
                    int newY = y + dy[i];
                    int newX = x + dx[i];

                    if(newY < 0 || newY >= N || newX < 0 || newX >= M) continue;
                    if(visited[newY, newX]) continue;
                    if(maps[newY, newX] == 0) continue;
                    if(newY == N-1 && newX == M-1)
                    {
                        answer = depth + 1;
                        break;
                    }

                    visited[newY, newX] = true;
                    q.Enqueue((newY, newX));
                }
                
                if(answer != -1) break;
            }
            
            if(answer != -1) break;
            depth++;
        }
        
        return answer;
    }
}